src/beamlime/applications/_nexus_helpers.py

```diff
@@ -1,155 +1,378 @@
+    )
+    # event_index - reference_time_index
+    # Increase event index according to the ``original_event_time_index``
+    event_index_dataset = find_nexus_structure(group, ("event_index",))
+    _set_values(
+        event_index_dataset,
+        np.concatenate(
+            (
+                _get_array_values(event_index_dataset),
+                data_piece["reference_time_index"] + len(original_event_time_offset),
             )
+        ),
+    )
+    # event_id - pixel_id
+    if data_piece.get("pixel_id") is not None:  # Pixel id is optional.
+        event_id_dataset = find_nexus_structure(group, ("event_id",))
+        _set_values(
+            event_id_dataset,
+            np.concatenate(
+                (_get_array_values(event_id_dataset), data_piece["pixel_id"])
+            ),
+        )
 
 
 def _node_name(n):
-    '''Defines the name of a nexus tree branch or leaf'''
-    config = n.get('config', {})
-    return n.get('name', config.get('name'))
+    """Defines the name of a nexus tree branch or leaf"""
+    config = n.get("config", {})
+    return n.get("name", config.get("name"))
 
 
 def iter_nexus_structure(
-    structure: Mapping, root: Optional[Tuple[Optional[str], ...]] = None
-) -> Iterable[Tuple[str, Mapping]]:
-    '''Visits all branches and leafs in the nexus tree'''
-    path = (*root, _node_name(structure)) if root is not None else ()
+    structure: NexusDictStructure, root: Optional[NexusPath] = None
+) -> Iterable[tuple[tuple[str | None, ...], Mapping]]:
+    """Visits all branches and leafs in the nexus tree"""
+    path = (*root, _node_name(structure)) if root is not None else tuple()
     yield path, structure
-    for child in structure.get('children', []):
+    for child in structure.get("children", []):
         yield from iter_nexus_structure(child, root=path)
 
 
 def find_nexus_structure(
-    structure: Mapping, path: Tuple[Optional[str], ...]
-) -> Optional[Mapping]:
-    '''Returns the branch or leaf associated with `path`, or None if not found'''
+    structure: NexusDictStructure, path: Sequence[Optional[str]]
+) -> Mapping:
+    """Returns the branch or leaf associated with `path`, or None if not found"""
     if len(path) == 0:
         return structure
     head, *tail = path
-    for child in structure['children']:
+    for child in structure["children"]:
         if head == _node_name(child):
             return find_nexus_structure(child, tail)
+    raise KeyError(f"Path {path} not found in the nexus structure.")
+
 
+def find_parent(
+    structure: NexusDictStructure, path: Sequence[Optional[str]]
+) -> NexusGroupDict:
+    # TODO: We can use typeguard here later.
+    return find_nexus_structure(structure, path[:-1])
 
-def path_to_message_group(structure: Mapping, message: Mapping) -> str:
-    '''Finds the group where the message should be merged'''
-    kind, content = message
+
+def find_ev44_matching_paths(
+    structure: NexusDictStructure, data_piece: DeserializedMessage
```

Comment:
(minor) Maybe `NexusTemplate` or `NexusStructure` instead of `NexusDictStructure`, the editor can tell us it is a dictionary already

Reply:
How would you know if it's a dictionary...?

Reply:
Ah you mean by the type? hmm okay we can keep the name simple.